lib/rhea-promise/index.ts

```diff
@@ -0,0 +1,662 @@
+  return new Promise<void>((resolve, reject) => {
+    if (!session || (session && typeof session !== "object")) {
+      reject(new Error("session is a required parameter and must be of tyepe 'object'."));
+    }
+
+    if (session.is_open()) {
+      function onClose(context: Context): void {
+        session.removeListener("session_close", onClose);
+        process.nextTick(() => {
+          debug("Resolving the promise as the amqp session has been closed.");
+          resolve();
+        });
+      }
+
+      function onError(context: Context): void {
+        session.removeListener("session_error", onError);
+        debug(`Error occurred while closing amqp session.`, context.session.error);
+        reject(context.session.error);
+      }
+
+      session.once("session_close", onClose);
+      session.once("session_error", onError);
+      session.close();
+    } else {
+      resolve();
+    }
+  });
+}
+
+/**
+ * Creates an amqp sender on the provided amqp session.
+ * @param {Session} session The amqp session object on which the sender link needs to be established.
+ * @param {SenderOptions} [options] Options that can be provided while creating an amqp sender.
+ * @return {Promise<Sender>} Promise<Sender>
+ * - **Resolves** the promise with the Sender object when rhea emits the "sender_open" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the "sender_close" event while trying
+ * to create an amqp sender.
+ */
+export async function createSender(session: any, options?: SenderOptions): Promise<any> {
+  return new Promise((resolve, reject) => {
+    if (!session || (session && typeof session !== "object")) {
+      reject(new Error("session is a required parameter and must be of tyepe 'object'."));
+    }
+
+    const sender = session.attach_sender(options);
+
+    function onOpen(context: any): void {
+      sender.removeListener("sendable", onOpen);
+      sender.removeListener("sender_close", onClose);
+      process.nextTick(() => {
+        debug(`Resolving the promise with amqp sender "${sender.name}".`);
+        resolve(sender);
+      });
+    }
+
+    function onClose(context: Context): void {
+      sender.removeListener("sendable", onOpen);
+      sender.removeListener("sender_close", onClose);
+      debug(`Error occurred while creating a sender over amqp connection.`, context.sender.error);
+      reject(context.sender.error);
+    }
+
+    sender.once("sendable", onOpen);
+    sender.once("sender_close", onClose);
+  });
+}
+
+/**
+ * Closes the amqp sender.
+ * @param {Sender} sender The amqp sender that needs to be closed.
+ * @return {Promise<void>} Promise<void>
+ * - **Resolves** the promise when rhea emits the "sender_close" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the
+ * "sender_error" event while trying to close an amqp sender.
+ */
+export async function closeSender(sender: any): Promise<void> {
```

Comment:
fwiw if you intend to split this into it's own library, i'd change the call pattern. for an external library it feels weird to have functions that take in the raw rhea objects. i think it makes more sense to provide wrappers around the objects which make the operations you're doing promise-based